### ISSUE TITLE
Add support for mongodb+srv. Add support for inserts to MongoDB.

### ISSUE THREAD

**Issue by allmazz** [ClickHouse/ClickHouse#69766]
## Context
In https://github.com/ClickHouse/ClickHouse/pull/63279 we introduced refactored MongoDB integration, but this is not yet completed.  This issue created to track the progress.

## mongodb+srv
[mongo-c-driver](https://github.com/ClickHouse/mongo-c-driver) requires modern `libresolv` with `ns_initparse` and `ns_parserr`.
So, we need simply add `libresolv` to all sysroots, what was successfully done, but...

We found a test what checks glib version, and locks it to `2.4`, which is too old(**2006** year).
https://github.com/ClickHouse/ClickHouse/blob/master/tests/ci/compatibility_check.py#L199-L202
https://s3.amazonaws.com/clickhouse-test-reports/63279/d802e2cb9dd07b60bc4edb26db5ba0fece43ecb2/compatibility_check__release_.html

All PRs was closed and reverted:
https://github.com/ClickHouse/sysroot/pull/27
https://github.com/ClickHouse/sysroot/pull/33
https://github.com/ClickHouse/sysroot/pull/34
https://github.com/ClickHouse/sysroot/pull/36

The lock should be reviewed and minimum-maximum versions should be updated to modern versions.

## Insert
Since address mapping is done, we can easily add inserts.
Inserts was supported by the legacy implementation, but was not documented(this is why we ignored this).

## use_legacy_mongodb_integration
Previous implementation is keeped, and can be enabled by the server setting `use_legacy_mongodb_integration`.
Currently this enabled by default.

After a few releases and feedback from users, previous implementation must be removed with the setting and the Poco driver implementation.

## Tasks
+ - [ ] https://github.com/ClickHouse/ClickHouse/issues/77911
+ - [ ] Add support for `mongodb+srv`
+ - [ ] Optional: add support for inserts

**Comment by rschu1ze** [comment]
in 24.12, both old and new MongoDB drivers are available but the new one was made default.
After the next LTS release (25.2?), the old driver should be removed. At that point no fallback will be possible anymore.

**Comment by allmazz** [comment]
@rschu1ze GitHub obsoletes checklists, so please convert https://github.com/ClickHouse/ClickHouse/issues/77911 to sub-issue

https://github.blog/changelog/2025-02-18-github-issues-projects-february-18th-update/#tasklist-blocks-will-be-retired-and-replaced-with-sub-issues

**Comment by rschu1ze** [comment]
@allmazz I don't mind if we use checklists or sub-issues for tracking.

As the author of the issue, you should be able to edit the original comment, right?

**Comment by rschu1ze** [comment]
This is the error I get _repeatedly_ when I try to "Convert to sub-issue": `Could not convert the checklist item to an issue. Please refresh the page and try again.`.

Anyways, the logistics (checklist, sub-item) really don't matter. I marked the first issue as complete now.

**Comment by LeckerenSirupwaffeln** [comment]
I'll make a PR for this, it's pretty easy to fix with c-areas, but need to re-factor MongoDBSource and Storage a bit as well since right now the connection is re-created on every query (mongocxx::client isn't saved anywhere), adding SRV on top of that will make it slow IMO (probs over 100 ms on every query of overhead just for connecting to client and verifying everything). SRV is designed with TTL limits in-mind anyway

**Comment by LeckerenSirupwaffeln** [comment]
Also mongocxx client supports SRV on it's own, but yeah that requires libresolv...

**Comment by allmazz** [comment]
I’ve already got some groundwork done on the MongoDBSource part, but libresolv...
Probably, we could integrate the driver with our resolver and cache, but it requires a lot of changes in it.

**Comment by LeckerenSirupwaffeln** [comment]
[This is WIP for solution without libresolv](https://github.com/ClickHouse/ClickHouse/compare/master...LeckerenSirupwaffeln:ClickHouse:feat_mongodb_srv), but if you want you could help me figure out if I am taking the right approach? I am trying to standardize access to c-ares, so I still need to change some things including making sure other parts of the codebase aren't using it "naked" anymore

So I standardized access to c-ares with _AsyncAresExecutor_, and then in Server.cpp I attached _AsyncAresExecutor_ to the global context background schedule pool, then created SRVResolver that can be used in local context like inside the _MongoDBStorage_

Does that make sense?

**Comment by allmazz** [comment]
Good start! You want to re-implement SRV support outside of the driver, right?
If yes, you can extend DNSResolver instead of creating new class, it's already has configurable updater(DNSCacheUpdater), and can be easily extended for TXT records. You don't need context here, because it's a singleton instance. One more benefit of using it is [a ready table to debug cache](https://github.com/ClickHouse/ClickHouse/pull/59856)

**Comment by LeckerenSirupwaffeln** [comment]
Yep, SRV support outside of libresolv to make it backwards compatible with the older operating systems, c-ares is supposed to work without libresolv for SRV

The DNS cache table is nice, If I extend DNSResolver, I guess it would just need its own resolveSRV(...) functionality to get hostname(s) and it would use the AsyncAresExecutor

I will see if it's easy to extend the DNSCache to work for SRV -> hostname(s), I also know that c-ares itself has caching internally, but I added another cache on top just to be able to give an almost expired hostname and resolve SRV again before it would expire

**Comment by allmazz** [comment]
Great! Feel free to tag me for review.
